contracts/BC_fusion/StakePool.sol

```diff
@@ -0,0 +1,248 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts-upgradeable/utils/structs/DoubleEndedQueueUpgradeable.sol";
+
+import "./StBNB.sol";
+import "./System.sol";
+
+interface IStakeHub {
+    function isPaused() external view returns (bool);
+    function getUnbondTime() external view returns (uint256);
+}
+
+contract StakePool is System, StBNB {
+    using DoubleEndedQueueUpgradeable for DoubleEndedQueueUpgradeable.Bytes32Deque;
+
+    uint256 public constant MAX_CLAIM_NUMBER = 20;
+
+    /*----------------- storage -----------------*/
+    address public validator;
+
+    uint256 private _totalReceivedReward; // just for statistics
+    uint256 private _totalPooledBNB; // total reward plus total BNB staked in the pool
+
+    // hash of the unbond request => unbond request
+    mapping(bytes32 => UnbondRequest) private _unbondRequests;
+    // user => unbond request queue(hash of the request)
+    mapping(address => DoubleEndedQueueUpgradeable.Bytes32Deque) private _unbondRequestsQueue;
+    // user => locked shares
+    mapping(address => uint256) private _lockedShares;
+    // user => personal unbond sequence
+    mapping(address => uint256) private _unbondSequence;
+
+    struct UnbondRequest {
+        uint256 sharesAmount;
+        uint256 bnbAmount;
+        uint256 unlockTime;
+    }
+
+    /*----------------- events -----------------*/
+    event Delegated(address indexed sender, uint256 sharesAmount, uint256 bnbAmount);
+    event RewardReceived(uint256 bnbAmount);
+    event UnbondRequested(address indexed sender, uint256 sharesAmount, uint256 bnbAmount, uint256 unlockTime);
+    event UnbondClaimed(address indexed sender, uint256 sharesAmount, uint256 bnbAmount);
+
+    /*----------------- modifiers -----------------*/
+    modifier whenNotPaused() {
+        require(!IStakeHub(STAKE_HUB_ADDR).isPaused(), "CONTRACT_IS_STOPPED");
+        _;
+    }
+
+    /*----------------- external functions -----------------*/
+    function initialize(address _validator) public payable initializer {
+        validator = _validator;
+
+        assert(msg.value != 0);
+        _bootstrapInitialHolder(msg.value);
+    }
+
+    function delegate(address _delegator) external payable onlyStakeHub whenNotPaused returns (uint256) {
```

Comment:
because this can only be called from `onlyStakeHub`, what about add the `whenNotPaused` on `onlyStakeHub` function, so that we dont need this modifier here.

Reply:
Done

Reply:
What is your consideration?